test/integration/volume/attach_detach_test.go

```diff
@@ -138,6 +144,183 @@ var defaultTimerConfig = attachdetach.TimerConfig{
+	}
+	gotNode, err := testClient.CoreV1().Nodes().UpdateStatus(context.TODO(), deepCopyNode, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to patch the node status : %v", err)
+	}
+	gotNode.Spec.Taints = append(gotNode.Spec.Taints, taint)
+	if _, err := testClient.CoreV1().Nodes().Update(context.TODO(), gotNode, metav1.UpdateOptions{}); err != nil {
+		t.Fatalf("Failed to taint the node : %v", err)
+	}
+
+	waitForNodeToBeNotReady(t, testClient, "node-sandbox")
+	waitForNodeToBeTainted(t, testClient, "node-sandbox", v1.TaintNodeOutOfService)
+	graceTime := int64(300)
+	err = testClient.CoreV1().Pods(namespaceName).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{
+		GracePeriodSeconds: &graceTime,
+	})
+	if err != nil {
+		t.Fatalf("error in deleting pod: %v", err)
+	}
+	waitForPodDeletionTimeStampToSet(t, testClient, pod.Name, namespaceName)
+	waitForMetric(t, metrics.ForceDetachMetricCounter.WithLabelValues(metrics.ForceDetachReasonOutOfService), 1, "detach-metrics")
+}
+
+// TestPodTerminationWithNodeOOSDetach integration test verifies that if `out-of-service` taint is applied to the node,
+// which is shutdown non gracefully, then all the pods will immediately get terminated and volume be immediately detached
+// without waiting for the default time out period.
+func TestPodTerminationWithNodeOOSDetach(t *testing.T) {
+	// Disable ServiceAccount admission plugin as we don't have serviceaccount controller running.
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{"--disable-admission-plugins=ServiceAccount"}, framework.SharedEtcd())
+	defer server.TearDownFn()
+
+	namespaceName := "test-volume-detach"
+	testClient, ctrl, pvCtrl, gcCtrl, informers := createAdClients(t, server, defaultSyncPeriod, defaultTimerConfig)
+	ns := framework.CreateNamespaceOrDie(testClient, namespaceName, t)
+	defer framework.DeleteNamespaceOrDie(testClient, ns, t)
+
+	// create a node
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "node-sandbox",
+			Annotations: map[string]string{
+				util.ControllerManagedAttachAnnotation: "true",
+			},
+		},
+	}
+	if _, err := testClient.CoreV1().Nodes().Create(context.TODO(), node, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to created node : %v", err)
+	}
+
+	// create fake pods with volumes
+	pod := fakePodWithVol(namespaceName)
+	if _, err := testClient.CoreV1().Pods(ns.Name).Create(context.TODO(), pod, metav1.CreateOptions{}); err != nil {
+		t.Errorf("Failed to create pod : %v", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	podInformer := informers.Core().V1().Pods().Informer()
+	// start the informer
+	informers.Start(ctx.Done())
+	informers.WaitForCacheSync(ctx.Done())
+
+	// run the controllers
+	go ctrl.Run(ctx)
+	go pvCtrl.Run(ctx)
+	go gcCtrl.Run(ctx)
+
+	waitToObservePods(t, podInformer, 1)
+	gotNode, err := testClient.CoreV1().Nodes().Get(context.TODO(), "node-sandbox", metav1.GetOptions{})
```

Comment:
test/integration/volume/attach_detach_test.go:276:2: SA4006: this value of `gotNode` is never used (staticcheck)
	gotNode, err := testClient.CoreV1().Nodes().Get(context.TODO(), "node-sandbox", metav1.GetOptions{})